### ISSUE TITLE
[cpumanager] AddContainer error: not enough cpus available to satisfy request

### ISSUE THREAD

**Issue by hex108** [kubernetes/kubernetes#79159]
**What happened**:
I created a `Job`,  and some pods encountered error `PreStartHookError`. Kubelet related logs are as following:

```
 I0603 06:30:00.131653   17205 policy_static.go:207] [cpumanager] allocateCpus: (numCPUs: 1)
 E0603 06:30:00.131661   17205 policy_static.go:187] [cpumanager] unable to allocate 1 CPUs (container id: 0ff6e7f99108df03f3c0539404111242212f4f07da7557164d52c41e3237c305, error: not enough cpus availab      le to satisfy request)
 E0603 06:30:00.131665   17205 cpu_manager.go:173] [cpumanager] AddContainer error: not enough cpus available to satisfy request
 E0603 06:30:00.131694   17205 kuberuntime_manager.go:775] container start failed: PreStartHookError: not enough cpus available to satisfy request
```

**What you expected to happen**:

There should be no `PreStartHookError` errors when running pod.

**How to reproduce it (as minimally and precisely as possible)**:

1. Run a local kubernetes cluster with latest kubernetes code, and use static cpu manager: 
   ```   
   $ hack/local-up-cluster.sh
   $ git diff 
    ...
       @@ -775,6 +775,8 @@ function start_kubelet {
           ${image_service_endpoint_args[@]+"${image_service_endpoint_args[@]}"}
           "--runtime-request-timeout=${RUNTIME_REQUEST_TIMEOUT}"
           "--port=${KUBELET_PORT}"
   +      "--cpu-manager-policy=static"
   +      "--system-reserved=cpu=1"
          ${KUBELET_FLAGS}
     )

   ```
2. Create a Job as following
   ```
   $ kubectl create -f ./job.yaml
   $  cat job.yaml
   apiVersion: batch/v1
   kind: Job
   metadata:
     name: nginx
   spec:
     parallelism: 6
     template:
       spec:
         containers:
         - name: nginx
           image: nginx:1.7.9
           command: ["/bin/ls",  "/do_not_exist_directory"]
           resources:
             requests:
               memory: "32Mi"
               cpu: "1"
             limits:
               memory: "32Mi"
               cpu: "1"
         restartPolicy: Never
   ```
   **Note**: If your local machine has too many cpu cores, you need increate `.spec.parallelism` to a larger value to reproduce the problem.
3. kubectl get pods, and you'll find some pods failed because of `PreStartHookError`.
  ```
  $ kubectl get pod
  NAME          READY   STATUS              RESTARTS   AGE
  nginx-44cz9   0/1     PreStartHookError   0          59s
nginx-7t952   0/1     Error               0          54s
nginx-9kmmh   0/1     PreStartHookError   0          59s
nginx-ftf64   0/1     Error               0          49s
nginx-kp7ww   0/1     PreStartHookError   0          44s
nginx-nz7h8   0/1     PreStartHookError   0          60s

  ```

**Anything else we need to know?**:

**Environment**: CentOS 3.10
- Kubernetes version (use `kubectl version`): latest code
- Cloud provider or hardware configuration: 
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by hex108** [comment]
Root cause: 

When pod's status is `PodFailed`, kubelet will try to kill containers in the pod at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1558, however these containers do not include finished containers(https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L811, https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/helpers.go#L232), so kubelet will not kill them, then cpu manager will not release their cpus. Although cpu manager will release them in its reconcile function `reconcileState`, there is some latency.

**Comment by jaypipes** [comment]
How many CPUs does your host node have? The static allocation policy means that you will need to have >6 CPUs on the node in order for this Job to run there (because you are reserving 1 CPU for the host via command line flags and you are using parallelism of 6). At least, that's what I think the static CPU manager resource allocation does....

**Comment by hex108** [comment]
> How many CPUs does your host node have? The static allocation policy means that you will need to have >6 CPUs on the node in order for this Job to run there (because you are reserving 1 CPU for the host via command line flags and you are using parallelism of 6). At least, that's what I think the static CPU manager resource allocation does....

There is 4 cpus. The pod of the job will finish in a while, it will not hold the cpu for a long time, and cpu manager will release their cpu once they finish.

**Comment by jaypipes** [comment]
@hex108 sorry for delayed response, Jun. I have some follow-up questions.

First, I would expect that one of your Pods would have a status of Succeeded, but I only see Pods in either Error or PreStartHookError status. Do you see Pods get to Succeeded state as well, or do all Pods that are launched for the Job always end up in either Error or PreStartHookError?

Secondly, have you tried reproducing this without the `--cpu-manager-policy=static` and `--system-reserved` flags and using a larger parallelism value?

Thanks!
-jay

**Comment by hex108** [comment]
> First, I would expect that one of your Pods would have a status of Succeeded, but I only see Pods in either Error or PreStartHookError status. Do you see Pods get to Succeeded state as well, or do all Pods that are launched for the Job always end up in either Error or PreStartHookError?

I make these pods fail intentionally to reproduce the bug. The pod has finished(failed actually), so scheduler could schedule a new pod to the node, however the kubelet node has not released the CPU that previous finished pod used, then the new pod failed to run because of not enough cpus available. It is a kubelet bug.

> 
> Secondly, have you tried reproducing this without the `--cpu-manager-policy=static` and `--system-reserved` flags and using a larger parallelism value?
> 

It is a bug of cpu manager static policy, so I think we might not need to test this case without static policy.

**Comment by jaypipes** [comment]
I have successfully reproduced this bug using this script:

```sh
#!/usr/bin/env

TESTNAME="repro-issue-79159"
TESTDIR=$(mktemp -d "/tmp/$TESTNAME.XXX")

echo "Using testdir: $TESTDIR"

cat > $TESTDIR/cluster-config.yaml <<EOF
# this config file contains all config fields with comments
kind: Cluster
apiVersion: kind.sigs.k8s.io/v1alpha3
# patch the generated kubeadm config with some extra settings
kubeadmConfigPatches:
- |
  apiVersion: kubeadm.k8s.io/v1beta2
  kind: JoinConfiguration
  metadata:
    name: config
  nodeRegistration:
    kubeletExtraArgs:
      "cpu-manager-policy": "static"
      "system-reserved": "cpu=1"
nodes:
- role: control-plane
- role: worker
EOF

if `kind get clusters | grep "$TESTNAME" >/dev/null 2>&1`; then
    echo "Stopping and removing worker node container ... "
    docker container stop $TESTNAME-worker >/dev/null 2>&1
    docker container rm $TESTNAME-worker >/dev/null 2>&1
    echo "Stopping and removing control plane node container ... "
    docker container stop $TESTNAME-control-plane >/dev/null 2>&1
    docker container rm $TESTNAME-control-plane >/dev/null 2>&1
    kind delete cluster --name "$TESTNAME"
fi

kind create cluster --name "$TESTNAME" --config $TESTDIR/cluster-config.yaml

sleep 3

kubectl --context kind-$TESTNAME get nodes

cat > $TESTDIR/job.yaml <<EOF
apiVersion: batch/v1
kind: Job
metadata:
  name: nginx-${TESTNAME}
spec:
  parallelism: 10
  template:
    spec:
      containers:
      - name: nginx
        image: nginx:1.7.9
        command: ["/bin/ls",  "/do_not_exist_directory"]
        resources:
          requests:
            memory: "32Mi"
            cpu: "1"
          limits:
            memory: "32Mi"
            cpu: "1"
      restartPolicy: Never
EOF

echo "Before creating job nginx-$TESTNAME, worker node looks like this:"
kubectl --context kind-$TESTNAME describe node $TESTNAME-worker

kubectl --context kind-$TESTNAME create -f $TESTDIR/job.yaml

echo "After creating job nginx-$TESTNAME, worker node looks like this:"
kubectl --context kind-$TESTNAME describe node $TESTNAME-worker
kubectl --context kind-$TESTNAME get pods -o wide -w
```

**NOTE**: my laptop has 8 cores which is why I set the `parallelism` to `10`

I see pods go into `PreStartHookError` eventually:

```
jaypipes@udb42383a8e5155:~/go/src/k8s.io/kubernetes$ kubectl get pods -o wide -w
NAME                            READY   STATUS              RESTARTS   AGE     IP            NODE                       NOMINATED NODE   READINESS GATES
nginx-repro-issue-79159-4mmxx   0/1     Error               0          6m57s   10.244.1.2    repro-issue-79159-worker   <none>           <none>
nginx-repro-issue-79159-77mjw   0/1     Error               0          6m57s   10.244.1.3    repro-issue-79159-worker   <none>           <none>
nginx-repro-issue-79159-bnkqm   0/1     Error               0          6m38s   10.244.1.10   repro-issue-79159-worker   <none>           <none>
nginx-repro-issue-79159-brhxm   0/1     Error               0          6m57s   10.244.1.4    repro-issue-79159-worker   <none>           <none>
nginx-repro-issue-79159-c65f9   0/1     Error               0          6m40s   10.244.1.9    repro-issue-79159-worker   <none>           <none>
nginx-repro-issue-79159-cj762   0/1     Error               0          6m57s   10.244.1.5    repro-issue-79159-worker   <none>           <none>
nginx-repro-issue-79159-mrrrq   0/1     PreStartHookError   0          6m57s   10.244.1.6    repro-issue-79159-worker   <none>           <none>
nginx-repro-issue-79159-pvb4z   0/1     PreStartHookError   0          6m36s   10.244.1.12   repro-issue-79159-worker   <none>           <none>
nginx-repro-issue-79159-r25kp   0/1     Error               0          6m57s   10.244.1.8    repro-issue-79159-worker   <none>           <none>
nginx-repro-issue-79159-sg57h   0/1     PreStartHookError   0          6m36s   10.244.1.13   repro-issue-79159-worker   <none>           <none>
nginx-repro-issue-79159-t99ft   0/1     Error               0          6m38s   10.244.1.11   repro-issue-79159-worker   <none>           <none>
```

but I note that the Node's `Allocatable.cpu` never gets decremented from `7` (which is number of cores minus the 1 system reserved core)

I will keep investigating where to place the fix for this, since I am still unsure that calling PostStartContainer in a couple places is actually fixing the root cause of the issue.

**Comment by dashpole** [comment]
Allocatable is effectively the "capacity for pods".  It isn't changed when pods are added or removed

**Comment by jaypipes** [comment]
> Allocatable is effectively the "capacity for pods". It isn't changed when pods are added or removed

How may I get the number of available CPU cores for a node?

**Comment by dashpole** [comment]
`kubectl describe no` will show the pods assigned to that node, and how much space is available.

**Comment by jaypipes** [comment]
For the static CPU core allocation, is the solution to add up the number of Pods that have a request/limit of 1 CPU then? Do you have a guess as to why this bug is popping up? It does seem like a race condition like @timstclair mentions in the related PR...

**Comment by dashpole** [comment]
Sadly I haven't kept up closely enough with the CPU manager design and implementation...

**Comment by jaypipes** [comment]
I think part of the problem -- at least from what I can tell -- is that dedicate CPU cores are accounted for in the same way as shared CPU quota. But dedicated CPUs and CPU shares are different things. One is a finite resource, the other is consumed in a quota/bucket-like accounting system.

But these things are treated as the same kind of resource, which means that there's no real way of determining the number of free dedicated cores that may be assigned to a container.

In other words, doing `k describe no` as you say above, leads to this:

```
Non-terminated Pods:         (7 in total)
  Namespace                  Name                             CPU Requests  CPU Limits  Memory Requests  Memory Limits  AGE
  ---------                  ----                             ------------  ----------  ---------------  -------------  ---
  default                    nginx-repro-issue-79159-6j4fx    1 (14%)       1 (14%)     32Mi (0%)        32Mi (0%)      27s
  default                    nginx-repro-issue-79159-b5pfp    1 (14%)       1 (14%)     32Mi (0%)        32Mi (0%)      1s
  default                    nginx-repro-issue-79159-blt7t    1 (14%)       1 (14%)     32Mi (0%)        32Mi (0%)      27s
  default                    nginx-repro-issue-79159-pdvg2    1 (14%)       1 (14%)     32Mi (0%)        32Mi (0%)      4s
  default                    nginx-repro-issue-79159-wgkbg    1 (14%)       1 (14%)     32Mi (0%)        32Mi (0%)      27s
  kube-system                kindnet-zbkvs                    100m (1%)     100m (1%)   50Mi (0%)        50Mi (0%)      31s
  kube-system                kube-proxy-mvstd                 0 (0%)        0 (0%)      0 (0%)           0 (0%)         31s
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests     Limits
  --------           --------     ------
  cpu                5100m (72%)  5100m (72%)
  memory             210Mi (1%)   210Mi (1%)
  ephemeral-storage  0 (0%)       0 (0%)
```

The "Allocated resources" section equates "1 CPU core (static) with "1000 millicores" but those things aren't the same type of resource IMHO.

**Comment by tallclair** [comment]
@jaypipes that makes sense. It seems like fixing it would be a pretty big change though, right?

**Comment by jaypipes** [comment]
> @jaypipes that makes sense. It seems like fixing it would be a pretty big change though, right?

Well, yes, definitely :) Definitely not proposing fixing the underlying issue in the solution for this bug. Still pondering how to identify where this race condition is actually occurring...

**Comment by riking** [comment]
Dupes are getting filed.

/remove-lifecycle rotten

**Comment by jgiannuzzi** [comment]
As per my [comments](https://github.com/kubernetes/kubernetes/issues/92436#issuecomment-653039945) in #92436, this issue has been fixed in v1.18.4 (via PR #90530) and v1.19.0-alpha.3 (via PR #90377).

I guess it can be closed?

**Comment by hex108** [comment]
Thanks @jgiannuzzi !
